docs(guides): add guide for integrating Tiptap into a mobile app (#877)
ueberdosis/tiptap-docs: src/content/guides/mobile-integration.mdx

---
tags:
  - type: editor
title: Integrate Tiptap into a mobile app
meta:
  title: Integrate Tiptap into a mobile app | Tiptap Editor Docs
  description: Run Tiptap inside a Flutter, iOS, or Android app and drive it from native code with a small JSON message protocol.
  category: Editor
---

import { Callout } from '@/components/ui/Callout'

> Tiptap runs in the browser, so to use it in a mobile app you render the editor in a webview and drive it from your native code. This guide shows a small JSON message protocol that lets your app run any editor command and receive any editor event, with the editor staying a normal Tiptap instance whose extensions render exactly as they do on the web. The same approach works in Flutter, iOS, and Android.

Reviewer: ```suggestion
To run Tiptap in a mobile or multi-platform app, render the editor in a WebView and drive it from your native code.

This guide shows a simple protocol that connects the multi-platform app to the Tiptap editor instance running in the WebView. This JSON-based protocol allows the app to send and receive any type of command or event. It works on any platform, including iOS, Android, React Native, and Flutter.
```

Re-wrote the intro. Changes: 
- Make it less personal (e.g. "you render")
- Clarify that it works on any platform.
- Remove redundant info (Tiptap runs on the browser)
- Clarify that this works on multi-platform apps, not only mobile apps
- Split and shortened the sentences.

    send({ type: 'response', id: command.id, ok: true })
    return
  }

  // The commands below run against the editor created here.
}
```

The `content` can be an HTML string or ProseMirror JSON: pass your saved document to open it, and set `editable: false` for a read-only view. To swap the document later (for example after fetching it from your API), send a `setContent` command that calls `editor.commands.setContent(content)`.

## Run any command

You don't need a bridge method per command. Send a generic `exec` message with a command name and its arguments, and run it on the editor's [command chain](/editor/api/commands). Add an `exec` branch to the `onCommand` from the previous step:

```json
{
  "type": "command",
  "id": "1",
  "name": "exec",
  "params": { "command": "toggleHeading", "args": [{ "level": 2 }] }
}
```

```js
// inside onCommand(command), after the init branch:
if (command.name === 'exec') {
  const { command: name, args = [] } = command.params
  const executed = editor
    .chain()
    .focus()
    [name](...args)
    .run()
  send({ type: 'response', id: command.id, ok: true, result: { executed } })
}
```

> `args` is a positional array spread into the command, so one path serves both object-argument commands (`toggleHeading([{ level: 2 }])`) and scalar ones (`setTextAlign(['left'])`). Every command in your build is reachable from native, with no glue per command.

Reviewer: The `args` explanation is misleading

This sentence says:

> object-argument commands (`toggleHeading([{ level: 2 }])`) and scalar ones (`setTextAlign(['left'])`)

Because the bridge spreads args, the actual calls are:
```
toggleHeading({ level: 2 })
setTextAlign('left')
```
The JSON representation is an array, but the invoked command arguments are not arrays. Suggested rewrite:

```suggestion
`args` is encoded as a positional array in JSON and spread into the command, so `{ "args": [{ "level": 2 }] }` calls `toggleHeading({ level: 2 })`, and `{ "args": ["left"] }` calls `setTextAlign("left")`.
```

  editor.on(name, () => {
    if (subscriptions.has(name)) {
      send({ type: 'event', name, payload: project(name) })
    }
  })
}
```

A raw event payload holds live objects (the `Editor`, a `Transaction`) that can't cross a JSON boundary, so forward a small projection instead, returning the selection for the document and selection events and an empty object for `focus` and `blur`:

```js
function project(name) {
  if (name === 'update' || name === 'selectionUpdate' || name === 'transaction') {
    const { from, to, empty } = editor.state.selection
    return { selection: { from, to, empty } }
  }
  return {} // focus, blur
}
```

The reference [`PROTOCOL.md`](https://github.com/ueberdosis/tiptap-mobile-bridge/blob/main/PROTOCOL.md) lists every forwarded event and its projected payload. When you need the full document after a change, ask for it with a command that returns `editor.getJSON()` or `editor.getHTML()`:

```js
// getJSON / getHTML are commands like any other: run, then reply with the document
if (command.name === 'getJSON') {
  send({ type: 'response', id: command.id, ok: true, result: editor.getJSON() })
}
```

## Drive a native toolbar

> A native toolbar needs to know which buttons are active (the selection is bold) and which are available (you can't make a list item bold). Compute that state when the selection changes and send it as a `selectionChanged` event, but only for the commands you actually render, never by looping over every command (see the warning below):

Reviewer: The example "you can't make a list item bold" is incorrect, you can make the content inside a list item bold.

Suggested replacement: whether a heading can be toggled at the current selection

int _nextId = 1;

// inbound: app to editor. Returns a future that completes with the response.
Future<Object?> send(String name, [Map<String, dynamic>? params]) {
  final id = '${_nextId++}';
  final completer = Completer<Object?>();
  _pending[id] = completer; // resolved when the response with this id arrives
  final command = {'type': 'command', 'id': id, 'name': name, if (params != null) 'params': params};
  // double-encode so the JSON is a valid JS string argument
  controller.evaluateJavascript(
    source: 'window.tiptapBridge.handleCommand(${jsonEncode(jsonEncode(command))})',
  );
  return completer.future;
}

// outbound: editor to app. Route responses, events, and host requests.
controller.addJavaScriptHandler(
  handlerName: 'tiptapBridge',
  callback: (args) {
    final message = jsonDecode(args.first as String) as Map<String, dynamic>;
    if (message['type'] == 'response') {
      final completer = _pending.remove(message['id']);
      message['ok'] == true
          ? completer?.complete(message['result'])
          : completer?.completeError(message['error']);
    }
    // 'event' and 'hostRequest' are handled here too (see tiptap_bridge.dart)
  },
);
```

> ### iOS (WKWebView)

Reviewer: ```suggestion
### iOS (`WKWebView`)
```

Add a script-message handler for outbound messages, and evaluate JavaScript for inbound commands:

```swift
// outbound: the editor calls window.webkit.messageHandlers.tiptapBridge.postMessage(json)
config.userContentController.add(self, name: "tiptapBridge")

// inbound
webView.evaluateJavaScript("window.tiptapBridge.handleCommand(\(jsStringLiteral))")
```

Send your first command only after `window.tiptapBridge` exists. `WKWebView`'s `didFinish` can fire before the editor's module installs it, so poll for it before sending that first command (see [`EditorWebView.swift`](https://github.com/ueberdosis/tiptap-mobile-bridge/blob/main/ios_example/TiptapBridgeExample/EditorWebView.swift) for the polling loop).

> ### Android (WebView)

Reviewer: ```suggestion
### Android (`WebView`)
```

---
tags:
  - type: editor

> title: Integrate Tiptap into a mobile app

Reviewer: Idea: simplify the title to `Tiptap for mobile apps`

This would mean changing the slug, the page meta title and the sidebar link, too

meta:
  title: Integrate Tiptap into a mobile app | Tiptap Editor Docs

>   description: Run Tiptap inside a Flutter, iOS, or Android app and drive it from native code with a small JSON message protocol.

Reviewer: ```suggestion
  description: Run Tiptap inside mobile and multi-platform apps like iOS, Android and Flutter. Control it from native code with a simple JSON message protocol.
```

Reviewer: Re-wrote the description
- Replace "small" with "simple".
- Make it not sound like only Flutter, iOS, or Android are supported.
- Mention mobile and multi-platform apps
